### PR TITLE
wip: add integration testing guidelines

### PR DESCRIPTION
**Pull request by KeisukeShima**
[upstreamへのPR作成前のディスカッション用です]

今度こそT4へPRを作成したはず！

変更内容
- interface testの説明を追加
- リファレンスを最下部へ移動
- 改行の位置を調整
- コード種類の不要な`{}`を削除

### CODE REVIEW

File: docs/contributing/testing-guidelines/integration-testing.md [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+which adds smoke test that ensures that node can be:
+
+1. launched with default parameter file,
+2. terminated with a standard `SIGTERM` signal,
+
+For full API documentation see [package design page](https://github.com/autowarefoundation/autoware.universe/blob/main/common/autoware_testing/design/autoware_testing-design.md).
+
+This API is not suitable for all smoke test cases. For example, it can not be used when some specific file location, like map, is required to be passed to the node or some preparation need to be conducted before node launch. In such cases use manual solution from [section below](#integration-test-with-a-single-node-component-test).
+
+### Interface tests
```
Comment (KeisukeShima) [inline docs/contributing/testing-guidelines/integration-testing.md:59]: この項目は新規追加しました。

File: docs/contributing/testing-guidelines/integration-testing.md [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+            callback=self._msg_received
+        )
+        self.addCleanup(self.node.destroy_subscription, sub)
+
+    def _msg_received(self, msg):
+        # Callback for ROS 2 subscriber used in the test
+        self.msgs.append(msg)
+
+    def get_message(self):
+        startlen = len(self.msgs)
```
Comment (kenji-miyake) [inline docs/contributing/testing-guidelines/integration-testing.md]:
```suggestion
        start_len = len(self.msgs)
```

File: docs/contributing/testing-guidelines/integration-testing.md [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+
+## Value of integration testing
+
+Integration tests determine if independently developed software modules work correctly when the modules are connected to each other. In ROS 2, the software modules are called nodes. As a special case, testing a single node can be referred to as component testing.
+
+Integration tests help to find the following types of errors:
+
+- Incompatible interaction between nodes, such as non-matching topics, different message types, or incompatible QoS settings
+- Reveal edge cases that were not touched with unit tests, such as a critical timing issue, network communication delay, disk I/O failure, and many other problems that can occur in production environments
+- Using tools like `stress` and `udpreplay`, performance of nodes is tested with real data or while the system is under high CPU/memory load, where situations such as `malloc` failures can be detected
```
Comment (kenji-miyake) [inline docs/contributing/testing-guidelines/integration-testing.md:17]: https://github.com/tier4/autoware-spell-check-dict/pull/245

File: docs/contributing/testing-guidelines/integration-testing.md [hunk 4]
```diff
@@ -0,0 +1,293 @@
+# Integration Testing
+
+## Introduction
```
Comment (kenji-miyake) [inline docs/contributing/testing-guidelines/integration-testing.md]: I think you can write an overview right below the title.

```suggestion
```

File: docs/contributing/testing-guidelines/integration-testing.md [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+This article motivates developers to adopt integration testing by explaining how to write, run, and evaluate the results of integration tests.
```
Comment (kenji-miyake) [inline docs/contributing/testing-guidelines/integration-testing.md]: Let's read and follow the Google developer documentation style guide as much as possible. :+1: 
https://developers.google.com/style/translation?hl=en#write-short,-clear,-and-precise-sentences

In this case, I think we can write like this:
```suggestion
This article explains how to write integration tests.
```

File: docs/contributing/testing-guidelines/integration-testing.md (hunk 1 above)
Comment (kenji-miyake) [reply docs/contributing/testing-guidelines/integration-testing.md:59]: それは良さそうですが、テストの全体像を先に書く必要がありそうです。
別のPRにありましたっけ？

File: docs/contributing/testing-guidelines/integration-testing.md [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+```python
+@launch_testing.post_shutdown_test()
+class TestProcessOutput(unittest.TestCase):
+
+    def test_exit_code(self, proc_output, proc_info, ndt_mapper):
+        # Check that process exits with code 0
+        launch_testing.asserts.assertExitCodes(proc_info, process=ndt_mapper)
+```
+
+## Running the test
```
Comment (kenji-miyake) [inline docs/contributing/testing-guidelines/integration-testing.md]: Let's move the descriptions about how to run tests to another page.
We need to keep pages short and simple.

**Review by kenji-miyake — COMMENTED** [review]
Let's discuss this PR tomorrow!

File: docs/contributing/testing-guidelines/integration-testing.md (hunk 1 above)
Comment (kenji-miyake) [reply docs/contributing/testing-guidelines/integration-testing.md:59]: （これか https://github.com/autowarefoundation/autoware.universe/pull/391）
こちらもautoware-documentationに持ってきた方が良さそうです。

File: docs/contributing/testing-guidelines/integration-testing.md (hunk 5 above)
Comment (kenji-miyake) [reply docs/contributing/testing-guidelines/integration-testing.md]: See also: https://github.com/autowarefoundation/autoware/pull/14
